src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+            //   - Since, by invariant, `self.ptr` addresses a range whose
+            //     length is not longer than `isize::MAX` bytes, so does `ptr`.
+            //   - Since, by invariant, `self.ptr` addresses a range which does
+            //     not wrap around the address space, so does `ptr`.
+            // - `validate_cast_and_convert_metadata` promises that the object
+            //   described by `split_at` is validly-aligned for `U`.
+            // - By invariant on `self`, `A` is guaranteed to live for at least
+            //   `'a`.
+            // - `U: 'a` by trait bound.
+            // - `ptr` conforms to the aliasing invariant of
+            //   `ALIASING_INVARIANT` because casting does not impact the
+            //   aliasing invariant
+            // - `ptr` conforms to the alignment invariant of `Aligned`, because
+            //   `validate_cast_and_convert_metadata` returns `None` if the cast
+            //   would leave `U` misaligned
+            // - `ptr`, trivially, conforms to the validity invariant of
+            //   `AnyValidity`
+            Some((unsafe { Ptr::new(ptr) }, split_at))
+        }
+
+        /// Attempts to cast `self` into a `U`, failing if all of the bytes of
+        /// `self` cannot be treated as a `U`.
+        ///
+        /// In particular, this method fails if `self` is not validly-aligned
+        /// for `U` or if `self`'s size is not a valid size for `U`.
+        ///
+        /// # Safety
+        ///
+        /// On success, the caller may assume that the returned pointer
+        /// references the same byte range as `self`.
+        #[allow(unused)]
+        #[inline(always)]
+        pub(crate) fn try_cast_into_no_leftover<U: 'a + ?Sized + KnownLayout>(
+            &self,
+        ) -> Option<
+            Ptr<'a, U, ALIASING_INVARIANT, { invariant::Aligned }, { invariant::AnyValidity }>,
+        > {
+            // TODO(#67): Remove this allow. See NonNulSlicelExt for more
+            // details.
+            #[allow(unstable_name_collisions)]
+            match self.try_cast_into(_CastType::_Prefix) {
+                Some((slf, split_at)) if split_at == self.len() => Some(slf),
+                Some(_) | None => None,
+            }
+        }
+    }
+}
+
+/// Projections through the referent.
+mod _project {
+    use super::*;
+
+    impl<
+            'a,
+            T,
+            const ALIASING_INVARIANT: invariant::Aliasing,
+            const ALIGNMENT_INVARIANT: invariant::Alignment,
+            const VALIDITY_INVARIANT: invariant::Validity,
+        > Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Projects a field from `self`.
+        ///
+        /// # Safety
+        ///
+        /// ## Preconditions
+        ///
+        /// The caller promises that `projector` projects a field of its
+        /// argument. Its argument will be `self` casted to a raw pointer.
```

Comment:
We probably also need to mention that it doesn't follow references or perform other indirection or whatever. Alternatively, we could require that the bytes referenced by `projector(p)` are a subset of those referenced by `p`.